posix: c_lib_ext: fnmatch: fix escape-oriented regression

cfriedt opened the pull request:
A regression in 936d0278bdfa773390469aa102a3de004f355514 introduced a subtle bug in the way that escaped expressions were handled.

The regression originated with the assumption that test data (originally adapted from a 3rd-party testsuite) was correct when it was in fact flawed.

Specifically, `fnmatch("[[?*\\]", "\\", 0)` should fail (`FNM_NOMATCH`), since the "\\" sequence (a single backslash after compilation) escapes the following ']' character, thus leaving the bracket expression incomplete.

As @keith-packard  has pointed out, [pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_13_01](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_13_01), a bracket expression is only interpreted as a bracket expression, when a proper bracket expression is formed.

Therefore, the pattern is interpreted as the sequence `'['`, `'['`, `'?'`, `'*'`, `']'` and the call should return `FNM_NOMATCH` to indicate failure rather than 0 to indicate success.

Additionally, add support for POSIX character classes, as that was a blocking change request. Partial implementation and tests pulled in from #97400.

Fixes #55186

Verified that tests are compatible with glibc.

Rebased after #99451 was merged

Commit:
posix: c_lib_ext: fnmatch: fix escape-oriented regression

A regression in 936d0278bdfa773390469aa102a3de004f355514 introduced a
subtle bug in the way that escaped expressions were handled.

The regression originated with the assumption that test data (originally
adapted from a 3rd-party testsuite) was correct when it was in fact
flawed.

Specifically, `fnmatch("[[?*\\]", "\\", 0)` should fail (`FNM_NOMATCH`),
since the "\\" sequence (a single backslash after compilation) escapes
the following ']' character, thus leaving the bracket expression
incomplete.

As @keith-packard  has pointed out,
https://pubs.opengroup.org/onlinepubs/9699919799/utilities/\
V3_chap02.html#tag_18_13_01 says that a bracket expression is only
interpreted as a bracket expression, when a proper bracket expression
is formed.

Therefore, the pattern is interpreted as the sequence
`'['`, `'['`, `'?'`, `*` (wildcard), `']'` and the call should return
`FNM_NOMATCH` to indicate failure rather than 0 to indicate success.

Added new test cases from #98827 and some commentary for subsequent
reviewers.

This change does not completely fix #55186 but is related to it.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented:
It would seem that even the glibc agrees with the former interpretation (that there is an incomplete bracket expression).

A sample program I wrote to test
```cpp
#include <fnmatch.h>
#include <stdio.h>

int main()
{
  const char *pat = "[abc\\]";
  const char *in = "a";
  int ret;

  ret = fnmatch(pat, in, 0);
  if (ret == 0) {
    printf("fnmatch(\"%s\", \"%s\", 0) -> success!\n", pat, in);
  } else {
    printf("fnmatch(\"%s\", \"%s\", 0) -> failed! (%d)\n", pat, in, ret);
  }

  return ret;
}
```
produces the output below.
```shell
$ /tmp/blah 
fnmatch("[abc\]", "a", 0) -> failed! (1)
```

It fails as well with `FNM_NOMATCH` when `pat = "[[?*\\]"` and `in = "\\"`.

keith-packard commented on tests/posix/c_lib_ext/src/fnmatch.c:
This test is also incorrect, again `"a"` matches `*[![:digit:]]*` while `"["` matches one of `[`, `:` or the range `d-d`.

I'd strongly prefer that the bugs in the Zephyr implementation be fixed before this series is merged; there's no particular reason to commit known-buggy code, especially as this function doesn't appear to be urgently in need of updates.

keith-packard reviewed (dismissed):
Please fix the remaining bugs in the implementation and update the tests to conform with the spec.

Commit:
posix: c_lib_ext: fnmatch: fix character class support

A couple of tests were inconsistent with glibc and picolibc.

Significant rework done to the `fnmatch()` implementation which included
refreshing that and the `rangematch()` implementations from commit

0a3b2e376d150258c8294c12a85bec99546ab84b

in https://github.com/lattera/freebsd

Removed `match_posix_class()` and implemented that functionality as
`rangematch_cc()`, which uses 64-bit integer comparison for matching
`[:alnum:]` et al instead of string comparison. That likely only works
for the "C" locale.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>
Signed-off-by: Harun Spago <harun.spago.code@gmail.com>

cfriedt commented:
The attached file can be used to verify that all of the tests in the testsuite are compatible with glibc as well.

```shell
gcc -o test-fnmatch-natively test-fnmatch-natively.c
./test-fnmatch-natively
```

[test-fnmatch-natively.c](https://github.com/user-attachments/files/23967609/test-fnmatch-natively.c)

Interestingly, the macOS libc is slightly incompatible.

```shell
test-fnmatch-natively.c:73: (fnmatch("[[?*\\]", "\\", 0)) failed (expected 1, actual 2)
test-fnmatch-natively.c:88: (fnmatch("[]?*\\]", "]", 0)) failed (expected 1, actual 0)
test-fnmatch-natively.c:115: ((fnmatch("[/b", "[/b", 0))) failed (expected 0, actual 2)
test-fnmatch-natively.c:138: ((fnmatch("*[![:digit:]]*/[![:d-d]", "a/b", 0x02))) failed (expected 0, actual 2)
test-fnmatch-natively.c:139: ((fnmatch("*[![:digit:]]*/[[:d-d]", "a/[", 0x02))) failed (expected 0, actual 2)
test-fnmatch-natively.c:140: (fnmatch("*[![:digit:]]*/[![:d-d]", "a/[", 0x02)) failed (expected 1, actual 2)
test-fnmatch-natively.c:189: (fnmatch("[[:foobarbaz:]]", "Z", 0)) failed (expected 1, actual 2)
```

Spago123 commented on lib/posix/c_lib_ext/fnmatch.c:
```diff
@@ -48,26 +53,121 @@
+	case FNM_CC5('b', 'l', 'a', 'n', 'k'):
+		ret = !isblank(ch);
+		break;
+	case FNM_CC5('c', 'n', 't', 'r', 'l'):
```
These part, would be more readable if the cases were indented, but it is fine this way also

cfriedt commented on tests/posix/c_lib_ext/src/fnmatch.c:
Fixed!
